sw/host/opentitanlib/src/otp/alert_handler.rs

```diff
@@ -0,0 +1,364 @@
+    class: [u32; ALERT_HANDLER_ALERT_CLASS_SHADOWED_MULTIREG_COUNT as usize],
+    /// ALERT_HANDLER_LOC_ALERT_REGWEN
+    loc_regwen: [u32; ALERT_HANDLER_LOC_ALERT_REGWEN_MULTIREG_COUNT as usize],
+    /// ALERT_HANDLER_LOC_ALERT_EN_SHADOWED
+    loc_en: [u32; ALERT_HANDLER_LOC_ALERT_EN_SHADOWED_MULTIREG_COUNT as usize],
+    /// ALERT_HANDLER_LOC_ALERT_CLASS_SHADOWED
+    loc_class: [u32; ALERT_HANDLER_LOC_ALERT_CLASS_SHADOWED_MULTIREG_COUNT as usize],
+    /// Alert handler class registers
+    class_regs: [AlertClassRegs; ALERT_HANDLER_PARAM_N_CLASSES as usize],
+}
+
+#[derive(TryFromPrimitive)]
+#[repr(u8)]
+enum AlertClass {
+    X = 0x94,
+    A = 0xee,
+    B = 0x64,
+    C = 0xa7,
+    D = 0x32,
+}
+
+impl AlertClass {
+    fn index(&self) -> usize {
+        match self {
+            AlertClass::A => 0,
+            AlertClass::B => 1,
+            AlertClass::C => 2,
+            AlertClass::D => 3,
+            AlertClass::X => 0,
+        }
+    }
+
+    fn from_index(index: usize) -> Self {
+        match index {
+            0 => AlertClass::A,
+            1 => AlertClass::B,
+            2 => AlertClass::C,
+            3 => AlertClass::D,
+            _ => AlertClass::X,
+        }
+    }
+}
+
+#[derive(TryFromPrimitive)]
+#[repr(u8)]
+enum AlertEnable {
+    None = 0xa9,
+    Enabled = 0x07,
+    Locked = 0xd2,
+}
+
+#[derive(TryFromPrimitive)]
+#[repr(u8)]
+enum AlertEscalate {
+    None = 0xd1,
+    Phase0 = 0xb9,
+    Phase1 = 0xcb,
+    Phase2 = 0x25,
+    Phase3 = 0x76,
+}
+
+struct AlertClassConfig {
+    enabled: AlertEnable,
+    escalate: AlertEscalate,
+    accum_thresh: u32,
+    timeout_cyc: u32,
+    phase_cycs: [u32; ALERT_HANDLER_PARAM_N_PHASES as usize],
+}
+
+impl Default for AlertClassRegs {
+    fn default() -> Self {
+        AlertClassRegs {
+            regwen: 1,
+            ctrl: 0,
+            accum_thresh: 0,
+            timeout_cyc: 0,
+            phase_cycs: [0; ALERT_HANDLER_PARAM_N_PHASES as usize],
+        }
+    }
+}
+
+impl Default for AlertRegs {
+    fn default() -> Self {
+        AlertRegs {
+            regwen: [1; ALERT_HANDLER_LOC_ALERT_REGWEN_MULTIREG_COUNT as usize],
+            loc_regwen: [1; ALERT_HANDLER_LOC_ALERT_REGWEN_MULTIREG_COUNT as usize],
+            en: [0; ALERT_HANDLER_ALERT_EN_SHADOWED_MULTIREG_COUNT as usize],
+            class: [0; ALERT_HANDLER_ALERT_CLASS_SHADOWED_MULTIREG_COUNT as usize],
+            loc_en: [0; ALERT_HANDLER_LOC_ALERT_EN_SHADOWED_MULTIREG_COUNT as usize],
+            loc_class: [0; ALERT_HANDLER_LOC_ALERT_CLASS_SHADOWED_MULTIREG_COUNT as usize],
+            class_regs: [Default::default(); ALERT_HANDLER_PARAM_N_CLASSES as usize],
+        }
+    }
+}
+
+impl AlertRegs {
+    pub fn crc32(self) -> u32 {
+        let crc = new_crc();
+        let mut digest = crc.digest();
+        self.crc32_add(&mut digest);
+        digest.finalize()
+    }
+
+    pub fn new<T: OtpRead>(lc_state: LcStateVal, otp: &T) -> Result<Self> {
```

Comment:
Thanks @jon-flatley! I know the mapping between the OTP values and alert_handler registers is far from trivial :)

TBH I didn't notice anything wrong but it's easy to miss a small error -- I think the best way to check this is to add e2e tests for several OTP configurations.

Reply:
Good suggestion, thanks @alphan. Please TAL at the additional test cases, one of which takes OTP values off the device to ensure it produces the same register values, and the other takes those register values and ensures that the correct CRC32 is produced.